add initial design overview and threat model

THS-on opened the pull request:
This add new section called design. This includes a more high level description of all the components in Keylime and a basic threat model.

THS-on commented:
> So validation the UEFI event log is more secure. You might want to add something about this to the text.

@edwards-n  I hinted to that in the overview, but I agree that it should be more explicit stated in the text.

aplanas reviewed (approved):
There are some minor typos that can be fixed later, but this test is really useful.

THS-on commented:
@edwards-n I added now a statement that the measured boot policy engine is preferred over static PCR values.

@aplanas If you have the time can you mark the typos? Then I can fix them in this PR.

@kkaarreell I added the inclusion of a nonce in the quote. Is this explanation enough?

aplanas commented:
>  If you have the time can you mark the typos? 

Sure, I am not English native and is usually others pointing me

aplanas commented on docs/design/threat_model.rst:
```diff
@@ -0,0 +1,66 @@
+
+From Keylime's perspective the core hardware like CPU, memory, motherboard is trusted, because it does not provide
+mechanisms to detect tampering with the hardware itself.
+Keylime chains its root of trust to the TPM therefore the TPM is deemed in general trustworthy. This trust is verified
```
"into the TPM"?

aplanas commented on docs/design/threat_model.rst:
```diff
@@ -0,0 +1,66 @@
+verified. The UEFI with Secure Boot enabled firmware and CRTM are generally trusted because it provides the UEFI event
+log and the API for other EFI applications to use the TPM. All the other applications in the boot chain are either
+measured by the firmware or the application that loads them (e.g. GRUB2 loads the kernel). The threat model does not
+require to trust arbitrary EFI applications during because it can be verified after boot what was executed.
```
"during the boot process"?

aplanas commented on docs/design/threat_model.rst:
```diff
@@ -0,0 +1,66 @@
+  * EK validation: The tenant allows Keylime to verify the EK certificate against the CAs of hardware manufacturers
+    or add custom validation steps. This is done to ensure that the EK belongs to an actual hardware TPM or a trusted
+    software TPM.
+  * AK enrollment: Using `MakeCredential`, `ActivateCredential` and enforcing certain key properties
```
Are "MakeCredential" and "ActivateCredential" TPM2 commands?

THS-on replied:
Yes

kkaarreell commented:
> @kkaarreell I added the inclusion of a nonce in the quote. Is this explanation enough?

@THS-on yes, thank you.

THS-on commented:
> @THS-on did you want to fix those typos in this PR or the next?

@mpeters I'll fix them tomorrow in this PR.

THS-on commented on docs/design/overview.rst:
```diff
@@ -0,0 +1,80 @@
+
+.. note::
+
+    The Keylime agent currently generates a AK on every startup and sends the EK and
```
>  IIRC the EK is not sent anywhere? You mean some public part? And is send to where?

Good catch. Yes I mean the public part. I added some clarification.

THS-on commented on docs/design/overview.rst:
```diff
@@ -0,0 +1,80 @@
+
+IMA validation
+""""""""""""""
+Keylime allows to verify files measured by IMA against either an provided allowlist or against a signature.
```
For signature verification see here: https://github.com/keylime/keylime/pull/718

I think the feature has not made it into the documentation yet.

THS-on replied:
The above PR was for automatically learning the keys. The feature for normal keyring feature is only documented here: https://keylime-docs.readthedocs.io/en/latest/rest_apis.html#post--v2.0-agents-agent_id-UUID.

I opened https://github.com/keylime/keylime-docs/issues/79 for adding better documentation on that.

THS-on commented on docs/design/overview.rst:
```diff
@@ -0,0 +1,80 @@
+Tenant
+------
+The tenant is a commandline management tool shipped by Keylime to manage agents.
+This includes adding and removing the agent from attestation, validation of the EK certificate against a cert store and
```
In this case it is a cert store.
